File: src/backend/workers/graph_builder/handler.py
"""Graph builder worker handler — transforms parse results into graph data."""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

import structlog
from ulid import ULID

from domains.artifacts.models import ArtifactError, ArtifactStatus
from domains.artifacts.repository import ArtifactRepository
from domains.graph.id_generation import generate_component_id, generate_edge_id
from domains.graph.repository import GraphRepository
from domains.projects.repository import ProjectRepository
from shared.cosmos import CosmosService
from shared.event_types import EVENT_GRAPH_BUILD_FAILED, EVENT_GRAPH_UPDATED
from shared.events import EventGridPublisher, build_cloud_event
from workers.base import PermanentError, TransientError, WorkerHandler

logger: structlog.stdlib.BoundLogger = structlog.get_logger(__name__)

DATABASE_NAME = "integration-copilot"
PROJECTS_CONTAINER = "projects"

# Statuses indicating the artifact has already progressed past graph building.
_POST_GRAPH_STATUSES: frozenset[ArtifactStatus] = frozenset(
    {
        ArtifactStatus.GRAPH_BUILT,
        ArtifactStatus.GRAPH_FAILED,
    }
)


class GraphBuilderHandler(WorkerHandler):
    """Process ``ArtifactParsed`` events by building graph data from parse results."""

    def __init__(
        self,
        artifact_repository: ArtifactRepository,
        graph_repository: GraphRepository,
        project_repository: ProjectRepository,
        cosmos_service: CosmosService,
        event_publisher: EventGridPublisher,
    ) -> None:
        self._artifact_repo = artifact_repository
        self._graph_repo = graph_repository
        self._project_repo = project_repository
        self._cosmos = cosmos_service
        self._publisher = event_publisher

    # -- WorkerHandler interface -----------------------------------------------

    async def is_already_processed(self, event_data: dict[str, Any]) -> bool:
        """Return ``True`` if the artifact has already been graph-built or failed."""
        tenant_id = event_data["tenantId"]
        artifact_id = event_data["artifactId"]

        artifact = await self._artifact_repo.get_by_id(tenant_id, artifact_id)
        if artifact is None:
            return False

        return artifact.status in _POST_GRAPH_STATUSES

    async def handle(self, event_data: dict[str, Any]) -> None:
        """Build graph from parse result: parsed → graph_building → graph_built."""
        tenant_id = event_data["tenantId"]
        project_id = event_data["projectId"]
        artifact_id = event_data["artifactId"]
        parse_result_id = event_data.get("parseResultId")

        log = logger.bind(
            tenant_id=tenant_id,
            project_id=project_id,
            artifact_id=artifact_id,
        )

        # parsed → graph_building
        try:
            await self._artifact_repo.update_status(
                tenant_id, artifact_id, ArtifactStatus.GRAPH_BUILDING
            )
        except Exception as exc:
            raise TransientError(f"Failed to transition to graph_building: {exc}") from exc

        # Load parse result from Cosmos DB
        parse_result = await self._load_parse_result(tenant_id, parse_result_id, artifact_id)
        if parse_result is None:
            raise PermanentError(
                f"Parse result not found for artifact {artifact_id}"
            )

        partition_key = f"{tenant_id}:{project_id}"
        now = datetime.now(UTC)

        # Get current project to determine graph version
        project = await self._project_repo.get_by_id(tenant_id, project_id)
        if project is None:
            raise PermanentError(f"Project {project_id} not found for tenant {tenant_id}")
        new_graph_version = project.graph_version + 1

        # Build temp_id → stable_id mapping
        id_map: dict[str, str] = {}

        # Process components
        components = parse_result.get("components", [])
        for comp in components:
            temp_id = comp.get("tempId", "")
            component_type = comp.get("componentType", "")
            name = comp.get("name", "")
            display_name = comp.get("displayName", name)

            stable_id = generate_component_id(tenant_id, project_id, component_type, name)
            id_map[temp_id] = stable_id

            doc = {
                "id": stable_id,
                "partitionKey": partition_key,
                "type": "component",
                "tenantId": tenant_id,
                "projectId": project_id,
                "artifactId": artifact_id,
                "componentType": component_type,
                "name": name,
                "displayName": display_name,
                "properties": comp.get("properties", {}),
                "tags": [],
                "graphVersion": new_graph_version,
                "createdAt": now.isoformat(),
                "updatedAt": now.isoformat(),
            }
            try:
                await self._graph_repo.upsert_component(doc)
            except Exception as exc:
                raise TransientError(f"Failed to upsert component {stable_id}: {exc}") from exc

        # Process external references as components
        external_refs = parse_result.get("externalReferences", [])
        for ref in external_refs:
            temp_id = ref.get("tempId", "")
            component_type = ref.get("componentType", "external_service")
            name = ref.get("name", "")
            display_name = ref.get("displayName", name)

            stable_id = generate_component_id(tenant_id, project_id, component_type, name)
            id_map[temp_id] = stable_id

            doc = {
                "id": stable_id,
                "partitionKey": partition_key,
                "type": "component",
                "tenantId": tenant_id,
                "projectId": project_id,
                "artifactId": artifact_id,
                "componentType": component_type,
                "name": name,
                "displayName": display_name,
                "properties": {"inferredFrom": ref.get("inferredFrom", "")},
                "tags": ["external"],
                "graphVersion": new_graph_version,
                "createdAt": now.isoformat(),
                "updatedAt": now.isoformat(),
            }
            try:
                await self._graph_repo.upsert_component(doc)
            except Exception as exc:
                raise TransientError(f"Failed to upsert external ref {stable_id}: {exc}") from exc

        # Process edges — resolve temp IDs to stable IDs
        edges = parse_result.get("edges", [])
        for edge in edges:
            source_temp_id = edge.get("sourceTempId", "")
            target_temp_id = edge.get("targetTempId", "")
            edge_type = edge.get("edgeType", "")

            source_id = id_map.get(source_temp_id, source_temp_id)
            target_id = id_map.get(target_temp_id, target_temp_id)

            edge_id = generate_edge_id(source_id, target_id, edge_type)
            doc = {
                "id": edge_id,
                "partitionKey": partition_key,
                "type": "edge",
                "tenantId": tenant_id,
                "projectId": project_id,
                "sourceComponentId": source_id,
                "targetComponentId": target_id,
                "edgeType": edge_type,
                "properties": edge.get("properties", {}) or {},
                "artifactId": artifact_id,
                "graphVersion": new_graph_version,
                "createdAt": now.isoformat(),
            }
            try:
                await self._graph_repo.upsert_edge(doc)
            except Exception as exc:
                raise TransientError(f"Failed to upsert edge {edge_id}: {exc}") from exc

        # Compute and upsert graph summary
        try:
            component_counts = await self._graph_repo.count_by_type(partition_key, "component")
            edge_counts = await self._graph_repo.count_by_type(partition_key, "edge")
            total_components = sum(component_counts.values())
            total_edges = sum(edge_counts.values())

            summary_id = f"gs_{partition_key}"
            summary_doc = {
                "id": summary_id,
                "partitionKey": partition_key,
                "type": "graph_summary",
                "tenantId": tenant_id,
                "projectId": project_id,
                "graphVersion": new_graph_version,
                "totalComponents": total_components,
                "totalEdges": total_edges,
                "componentCounts": component_counts,
                "edgeCounts": edge_counts,
                "updatedAt": now.isoformat(),
            }
            await self._graph_repo.upsert_summary(summary_doc)
        except Exception as exc:
            raise TransientError(f"Failed to compute/upsert graph summary: {exc}") from exc

        # Increment project graphVersion
        try:
            project.graph_version = new_graph_version
            await self._project_repo.update(project)
        except Exception as exc:
            raise TransientError(f"Failed to increment project graphVersion: {exc}") from exc

        # graph_building → graph_built
        try:
            await self._artifact_repo.update_status(
                tenant_id, artifact_id, ArtifactStatus.GRAPH_BUILT
            )
        except Exception as exc:
            raise TransientError(f"Failed to transition to graph_built: {exc}") from exc

        # Publish GraphUpdated event
        event = build_cloud_event(
            event_type=EVENT_GRAPH_UPDATED,
            source="/integration-copilot/worker/graph-builder",
            subject=f"tenants/{tenant_id}/projects/{project_id}",
            data={
                "tenantId": tenant_id,
                "projectId": project_id,
                "artifactId": artifact_id,
                "graphVersion": new_graph_version,
                "totalComponents": total_components,
                "totalEdges": total_edges,
            },
        )
        await self._publisher.publish_event(event)
        log.info(
            "graph_built",
            graph_version=new_graph_version,
            components=total_components,
            edges=total_edges,
        )

    async def handle_failure(self, event_data: dict[str, Any], error: Exception) -> None:
        """Transition artifact to ``graph_failed`` on permanent error."""
        tenant_id = event_data["tenantId"]
        artifact_id = event_data["artifactId"]

        try:
            artifact = await self._artifact_repo.get_by_id(tenant_id, artifact_id)
            if artifact is not None and artifact.status not in _POST_GRAPH_STATUSES:
                try:
                    await self._artifact_repo.update_status(
                        tenant_id, artifact_id, ArtifactStatus.GRAPH_FAILED
                    )
                except Exception:
                    artifact.error = ArtifactError(
                        code="GRAPH_BUILD_FAILED",
                        message=str(error),
                        occurredAt=datetime.now(UTC),
                    )
                    await self._artifact_repo.update(artifact)

            # Publish GraphBuildFailed event
            event = build_cloud_event(
                event_type=EVENT_GRAPH_BUILD_FAILED,
                source="/integration-copilot/worker/graph-builder",
                subject=(
                    f"tenants/{tenant_id}/projects/{event_data.get('projectId', 'unknown')}"
                    f"/artifacts/{artifact_id}"
                ),
                data={
                    "tenantId": tenant_id,
                    "projectId": event_data.get("projectId"),
                    "artifactId": artifact_id,
                    "error": str(error),
                },
            )
            await self._publisher.publish_event(event)
        except Exception:
            logger.error("handle_failure_error", artifact_id=artifact_id, exc_info=True)

    # -- Private helpers -------------------------------------------------------

    async def _load_parse_result(
        self,
        tenant_id: str,
        parse_result_id: str | None,
        artifact_id: str,
    ) -> dict | None:
        """Load parse result from the projects container in Cosmos DB."""
        container = await self._cosmos.get_container(DATABASE_NAME, PROJECTS_CONTAINER)

        # Try by parse_result_id first
        if parse_result_id:
            try:
                doc = await container.read_item(item=parse_result_id, partition_key=tenant_id)
                if doc.get("type") == "parse_result":
                    return doc
            except Exception:
                logger.warning("parse_result_not_found_by_id", parse_result_id=parse_result_id)

        # Fallback: query by artifactId
        query = (
            "SELECT * FROM c WHERE c.partitionKey = @tenantId "
            "AND c.type = 'parse_result' AND c.artifactId = @artifactId "
            "ORDER BY c.parsedAt DESC OFFSET 0 LIMIT 1"
        )
        params = [
            {"name": "@tenantId", "value": tenant_id},
            {"name": "@artifactId", "value": artifact_id},
        ]
        async for item in container.query_items(query=query, parameters=params):
            return item

        return None

File: src/worker-graph-builder-go/internal/worker/handler.go
// Package worker contains the graph builder domain handler.
package worker

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"time"

	"github.com/christopherhouse/integrisight/worker-graph-builder-go/internal/azure/cosmos"
	"github.com/christopherhouse/integrisight/worker-graph-builder-go/internal/azure/eventgrid"
	"github.com/christopherhouse/integrisight/worker-graph-builder-go/internal/graph"
)

// Event type constants.
const (
	EventArtifactParsed   = "com.integration-copilot.artifact.parsed.v1"
	EventGraphUpdated     = "com.integration-copilot.graph.updated.v1"
	EventGraphBuildFailed = "com.integration-copilot.graph.build-failed.v1"
)

// Cosmos DB constants.
const (
	DatabaseName      = "integration-copilot"
	ProjectsContainer = "projects"
	GraphContainer    = "graph"
	SubscriptionName  = "graph-builder"
	EventSourceGraph  = "/integration-copilot/worker/graph-builder"
)

// postGraphStatuses are artifact statuses indicating the artifact has already
// progressed past graph building.
var postGraphStatuses = map[string]bool{
	"graph_built":  true,
	"graph_failed": true,
}

// CosmosClient abstracts Cosmos DB operations (enables testing via mocks).
type CosmosClient interface {
	ReadItem(ctx context.Context, database, container, id, partitionKey string) (map[string]any, error)
	CreateItem(ctx context.Context, database, container, partitionKey string, document map[string]any) (map[string]any, error)
	ReplaceItem(ctx context.Context, database, container, id, partitionKey string, document map[string]any, etag string) (map[string]any, error)
	UpsertItem(ctx context.Context, database, container, partitionKey string, document map[string]any) (map[string]any, error)
	QueryItems(ctx context.Context, database, container, partitionKey, query string, params []cosmos.QueryParam) ([]map[string]any, error)
	ExecuteStoredProcedure(ctx context.Context, database, container, partitionKey, sprocName string, params []any) (map[string]any, error)
}

// Publisher abstracts event publishing (enables testing via mocks).
type Publisher interface {
	PublishEvent(ctx context.Context, event eventgrid.CloudEventOut) error
}

// GraphBuilderHandler processes ArtifactParsed events.
type GraphBuilderHandler struct {
	cosmos    CosmosClient
	publisher Publisher
}

// NewGraphBuilderHandler creates a GraphBuilderHandler.
func NewGraphBuilderHandler(cosmosClient CosmosClient, publisher Publisher) *GraphBuilderHandler {
	return &GraphBuilderHandler{
		cosmos:    cosmosClient,
		publisher: publisher,
	}
}

// AcceptedEventTypes returns the set of event types this handler accepts.
func (h *GraphBuilderHandler) AcceptedEventTypes() map[string]bool {
	return map[string]bool{EventArtifactParsed: true}
}

// IsAlreadyProcessed returns true if the artifact has already been graph-built or failed.
func (h *GraphBuilderHandler) IsAlreadyProcessed(ctx context.Context, eventData map[string]any) (bool, error) {
	tenantID, _ := eventData["tenantId"].(string)
	artifactID, _ := eventData["artifactId"].(string)

	artifact, err := h.cosmos.ReadItem(ctx, DatabaseName, ProjectsContainer, artifactID, tenantID)
	if err != nil {
		return false, fmt.Errorf("read artifact for idempotency: %w", err)
	}
	if artifact == nil {
		return false, nil
	}

	status, _ := artifact["status"].(string)
	return postGraphStatuses[status], nil
}

// Handle processes an ArtifactParsed event end-to-end.
func (h *GraphBuilderHandler) Handle(ctx context.Context, eventData map[string]any) error {
	tenantID, _ := eventData["tenantId"].(string)
	projectID, _ := eventData["projectId"].(string)
	artifactID, _ := eventData["artifactId"].(string)
	parseResultID, _ := eventData["parseResultId"].(string)

	if tenantID == "" {
		return Permanent("missing tenantId")
	}
	if projectID == "" {
		return Permanent("missing projectId")
	}
	if artifactID == "" {
		return Permanent("missing artifactId")
	}

	slog.Info("handle_graph_build", "tenant_id", tenantID, "project_id", projectID, "artifact_id", artifactID)

	// Step 1: Transition artifact → graph_building.
	if err := h.updateArtifactStatus(ctx, tenantID, artifactID, "graph_building"); err != nil {
		return Transient(fmt.Sprintf("failed to transition to graph_building: %v", err))
	}

	// Step 2: Load parse result.
	parseResult, err := h.loadParseResult(ctx, tenantID, parseResultID, artifactID)
	if err != nil {
		return Transient(fmt.Sprintf("failed to load parse result: %v", err))
	}
	if parseResult == nil {
		return Permanent(fmt.Sprintf("parse result not found for artifact %s", artifactID))
	}

	// Step 3: Load project for graphVersion.
	project, err := h.cosmos.ReadItem(ctx, DatabaseName, ProjectsContainer, projectID, tenantID)
	if err != nil {
		return Transient(fmt.Sprintf("failed to read project %s: %v", projectID, err))
	}
	if project == nil {
		return Permanent(fmt.Sprintf("project %s not found for tenant %s", projectID, tenantID))
	}

	currentGraphVersion := toInt64(project["graphVersion"])
	newGraphVersion := currentGraphVersion + 1
	partitionKey := tenantID + ":" + projectID
	now := time.Now().UTC()
	nowStr := now.Format(time.RFC3339)

	// Step 4–6: Build id mapping and process components/edges.
	idMap := make(map[string]string)

	// Process components.
	components := toSlice(parseResult["components"])
	for _, raw := range components {
		comp := toMap(raw)
		if comp == nil {
			continue
		}
		tempID, _ := comp["tempId"].(string)
		componentType, _ := comp["componentType"].(string)
		name, _ := comp["name"].(string)
		displayName, _ := comp["displayName"].(string)
		if displayName == "" {
			displayName = name
		}

		stableID := graph.GenerateComponentID(tenantID, projectID, componentType, name)
		if tempID != "" {
			idMap[tempID] = stableID
		}

		properties := toMapAny(comp["properties"])
		doc := map[string]any{
			"id":            stableID,
			"partitionKey":  partitionKey,
			"type":          "component",
			"tenantId":      tenantID,
			"projectId":     projectID,
			"artifactId":    artifactID,
			"componentType": componentType,
			"name":          name,
			"displayName":   displayName,
			"properties":    properties,
			"tags":          []any{},
			"graphVersion":  newGraphVersion,
			"createdAt":     nowStr,
			"updatedAt":     nowStr,
		}
		if _, err := h.cosmos.UpsertItem(ctx, DatabaseName, GraphContainer, partitionKey, doc); err != nil {
			return Transient(fmt.Sprintf("failed to upsert component %s: %v", stableID, err))
		}
	}

	// Process external references.
	externalRefs := toSlice(parseResult["externalReferences"])
	for _, raw := range externalRefs {
		ref := toMap(raw)
		if ref == nil {
			continue
		}
		tempID, _ := ref["tempId"].(string)
		componentType, _ := ref["componentType"].(string)
		if componentType == "" {
			componentType = "external_service"
		}
		name, _ := ref["name"].(string)
		displayName, _ := ref["displayName"].(string)
		if displayName == "" {
			displayName = name
		}
		inferredFrom, _ := ref["inferredFrom"].(string)

		stableID := graph.GenerateComponentID(tenantID, projectID, componentType, name)
		if tempID != "" {
			idMap[tempID] = stableID
		}

		doc := map[string]any{
			"id":            stableID,
			"partitionKey":  partitionKey,
			"type":          "component",
			"tenantId":      tenantID,
			"projectId":     projectID,
			"artifactId":    artifactID,
			"componentType": componentType,
			"name":          name,
			"displayName":   displayName,
			"properties":    map[string]any{"inferredFrom": inferredFrom},
			"tags":          []any{"external"},
			"graphVersion":  newGraphVersion,
			"createdAt":     nowStr,
			"updatedAt":     nowStr,
		}
		if _, err := h.cosmos.UpsertItem(ctx, DatabaseName, GraphContainer, partitionKey, doc); err != nil {
			return Transient(fmt.Sprintf("failed to upsert external ref %s: %v", stableID, err))
		}
	}

	// Process edges.
	edges := toSlice(parseResult["edges"])
	for _, raw := range edges {
		edge := toMap(raw)
		if edge == nil {
			continue
		}
		sourceTempID, _ := edge["sourceTempId"].(string)
		targetTempID, _ := edge["targetTempId"].(string)
		edgeType, _ := edge["edgeType"].(string)

		sourceID := idMap[sourceTempID]
		if sourceID == "" {
			sourceID = sourceTempID
		}
		targetID := idMap[targetTempID]
		if targetID == "" {
			targetID = targetTempID
		}

		edgeID := graph.GenerateEdgeID(sourceID, targetID, edgeType)
		edgeProps := toMapAny(edge["properties"])

		doc := map[string]any{
			"id":               edgeID,
			"partitionKey":     partitionKey,
			"type":             "edge",
			"tenantId":         tenantID,
			"projectId":        projectID,
			"sourceComponentId": sourceID,
			"targetComponentId": targetID,
			"edgeType":         edgeType,
			"properties":       edgeProps,
			"artifactId":       artifactID,
			"graphVersion":     newGraphVersion,
			"createdAt":        nowStr,
		}
		if _, err := h.cosmos.UpsertItem(ctx, DatabaseName, GraphContainer, partitionKey, doc); err != nil {
			return Transient(fmt.Sprintf("failed to upsert edge %s: %v", edgeID, err))
		}
	}

	// Step 10: Compute graph summary via stored procedure.
	summaryResult, err := h.cosmos.ExecuteStoredProcedure(ctx, DatabaseName, GraphContainer, partitionKey, "graphCountByTypes", nil)
	if err != nil {
		return Transient(fmt.Sprintf("failed to execute graphCountByTypes sproc: %v", err))
	}

	componentCounts := toMapAny(summaryResult["componentCounts"])
	edgeCounts := toMapAny(summaryResult["edgeCounts"])
	totalComponents := toInt64(summaryResult["totalComponents"])
	totalEdges := toInt64(summaryResult["totalEdges"])

	summaryID := "gs_" + partitionKey
	summaryDoc := map[string]any{
		"id":              summaryID,
		"partitionKey":    partitionKey,
		"type":            "graph_summary",
		"tenantId":        tenantID,
		"projectId":       projectID,
		"graphVersion":    newGraphVersion,
		"totalComponents": totalComponents,
		"totalEdges":      totalEdges,
		"componentCounts": componentCounts,
		"edgeCounts":      edgeCounts,
		"updatedAt":       nowStr,
	}
	if _, err := h.cosmos.UpsertItem(ctx, DatabaseName, GraphContainer, partitionKey, summaryDoc); err != nil {
		return Transient(fmt.Sprintf("failed to upsert graph summary: %v", err))
	}

	// Step 11: Increment project graphVersion.
	if err := h.incrementProjectGraphVersion(ctx, tenantID, projectID, newGraphVersion); err != nil {
		return Transient(fmt.Sprintf("failed to increment project graphVersion: %v", err))
	}

	// Step 12: Transition artifact → graph_built.
	if err := h.updateArtifactStatus(ctx, tenantID, artifactID, "graph_built"); err != nil {
		return Transient(fmt.Sprintf("failed to transition to graph_built: %v", err))
	}

	// Step 13: Publish GraphUpdated event.
	event := eventgrid.BuildCloudEvent(
		EventGraphUpdated,
		EventSourceGraph,
		fmt.Sprintf("tenants/%s/projects/%s", tenantID, projectID),
		map[string]any{
			"tenantId":        tenantID,
			"projectId":       projectID,
			"artifactId":      artifactID,
			"graphVersion":    newGraphVersion,
			"totalComponents": totalComponents,
			"totalEdges":      totalEdges,
		},
	)
	if err := h.publisher.PublishEvent(ctx, event); err != nil {
		slog.Warn("graph_updated_event_publish_failed", "error", err)
	}

	slog.Info("graph_built",
		"tenant_id", tenantID,
		"project_id", projectID,
		"graph_version", newGraphVersion,
		"components", totalComponents,
		"edges", totalEdges,
	)
	return nil
}

// HandleFailure transitions the artifact to graph_failed and publishes a GraphBuildFailed event.
func (h *GraphBuilderHandler) HandleFailure(ctx context.Context, eventData map[string]any, errMsg string) {
	tenantID, _ := eventData["tenantId"].(string)
	artifactID, _ := eventData["artifactId"].(string)
	projectID, _ := eventData["projectId"].(string)

	artifact, readErr := h.cosmos.ReadItem(ctx, DatabaseName, ProjectsContainer, artifactID, tenantID)
	if readErr == nil && artifact != nil {
		status, _ := artifact["status"].(string)
		if !postGraphStatuses[status] {
			if err := h.updateArtifactStatus(ctx, tenantID, artifactID, "graph_failed"); err != nil {
				slog.Error("update_status_graph_failed_error", "artifact_id", artifactID, "error", err)
			}
		}
	}

	event := eventgrid.BuildCloudEvent(
		EventGraphBuildFailed,
		EventSourceGraph,
		fmt.Sprintf("tenants/%s/projects/%s/artifacts/%s", tenantID, projectID, artifactID),
		map[string]any{
			"tenantId":   tenantID,
			"projectId":  projectID,
			"artifactId": artifactID,
			"error":      errMsg,
		},
	)
	if err := h.publisher.PublishEvent(ctx, event); err != nil {
		slog.Error("graph_build_failed_event_publish_error", "artifact_id", artifactID, "error", err)
	}
}

// --- Private helpers ---

// updateArtifactStatus performs a read-then-replace with ETag optimistic concurrency.
func (h *GraphBuilderHandler) updateArtifactStatus(ctx context.Context, tenantID, artifactID, newStatus string) error {
	doc, err := h.cosmos.ReadItem(ctx, DatabaseName, ProjectsContainer, artifactID, tenantID)
	if err != nil {
		return fmt.Errorf("read artifact: %w", err)
	}
	if doc == nil {
		return fmt.Errorf("artifact %s not found for tenant %s", artifactID, tenantID)
	}

	etag, _ := doc["_etag"].(string)
	doc["status"] = newStatus
	doc["updatedAt"] = time.Now().UTC().Format(time.RFC3339)

	if _, err := h.cosmos.ReplaceItem(ctx, DatabaseName, ProjectsContainer, artifactID, tenantID, doc, etag); err != nil {
		return fmt.Errorf("replace artifact: %w", err)
	}
	return nil
}

// incrementProjectGraphVersion reads the project and sets graphVersion = newVersion.
func (h *GraphBuilderHandler) incrementProjectGraphVersion(ctx context.Context, tenantID, projectID string, newVersion int64) error {
	doc, err := h.cosmos.ReadItem(ctx, DatabaseName, ProjectsContainer, projectID, tenantID)
	if err != nil {
		return fmt.Errorf("read project: %w", err)
	}
	if doc == nil {
		return fmt.Errorf("project %s not found for tenant %s", projectID, tenantID)
	}

	etag, _ := doc["_etag"].(string)
	doc["graphVersion"] = newVersion
	doc["updatedAt"] = time.Now().UTC().Format(time.RFC3339)

	if _, err := h.cosmos.ReplaceItem(ctx, DatabaseName, ProjectsContainer, projectID, tenantID, doc, etag); err != nil {
		return fmt.Errorf("replace project: %w", err)
	}
	return nil
}

// loadParseResult tries to load by parseResultID first, then falls back to
// querying by artifactID (most recent parse result, ORDER BY parsedAt DESC).
func (h *GraphBuilderHandler) loadParseResult(ctx context.Context, tenantID, parseResultID, artifactID string) (map[string]any, error) {
	if parseResultID != "" {
		doc, err := h.cosmos.ReadItem(ctx, DatabaseName, ProjectsContainer, parseResultID, tenantID)
		if err == nil && doc != nil {
			if docType, _ := doc["type"].(string); docType == "parse_result" {
				return doc, nil
			}
		}
		slog.Warn("parse_result_not_found_by_id", "parse_result_id", parseResultID)
	}

	// Fallback: query by artifactId.
	query := "SELECT * FROM c WHERE c.partitionKey = @tenantId AND c.type = 'parse_result' AND c.artifactId = @artifactId ORDER BY c.parsedAt DESC OFFSET 0 LIMIT 1"
	params := []cosmos.QueryParam{
		{Name: "@tenantId", Value: tenantID},
		{Name: "@artifactId", Value: artifactID},
	}
	docs, err := h.cosmos.QueryItems(ctx, DatabaseName, ProjectsContainer, tenantID, query, params)
	if err != nil {
		return nil, fmt.Errorf("query parse result: %w", err)
	}
	if len(docs) == 0 {
		return nil, nil
	}
	return docs[0], nil
}

// --- JSON coercion helpers ---

func toSlice(v any) []any {
	if v == nil {
		return nil
	}
	s, _ := v.([]any)
	return s
}

func toMap(v any) map[string]any {
	if v == nil {
		return nil
	}
	m, _ := v.(map[string]any)
	return m
}

func toMapAny(v any) map[string]any {
	if m, ok := v.(map[string]any); ok {
		return m
	}
	// Try JSON round-trip for edge cases (e.g. json.Number).
	if v != nil {
		raw, err := json.Marshal(v)
		if err == nil {
			var out map[string]any
			if json.Unmarshal(raw, &out) == nil {
				return out
			}
		}
	}
	return map[string]any{}
}

func toInt64(v any) int64 {
	switch val := v.(type) {
	case int64:
		return val
	case float64:
		return int64(val)
	case json.Number:
		n, _ := val.Int64()
		return n
	case int:
		return int64(val)
	}
	return 0
}
